File: DFS/623. Add One Row to Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        if depth == 2:
            root.left = TreeNode(val, root.left, None)
            root.right = TreeNode(val, None, root.right)
        elif depth == 1:
            return TreeNode(val, root, None)
        else:
            if root.left:
                root.left = self.addOneRow(root.left, val, depth-1)
            if root.right:
                root.right = self.addOneRow(root.right, val, depth-1)
        return root

File: DFS/623. Add One Row to Tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* addOneRow(struct TreeNode* root, int val, int depth){
    if (depth == 1){
        struct TreeNode* node = (struct TreeNode*)malloc(sizeof(struct TreeNode));
        node->val = val;
        node->left = root;
        node->right = NULL;
        return node;
    }
    else if (depth == 2){
        struct TreeNode* node_1 = (struct TreeNode*)malloc(sizeof(struct TreeNode));
        node_1->val = val;
        node_1->left = root->left;
        node_1->right = NULL;
        root->left = node_1;
        
        struct TreeNode* node_2 = (struct TreeNode*)malloc(sizeof(struct TreeNode));
        node_2->val = val;
        node_2->right = root->right;
        node_2->left = NULL;
        root->right = node_2;
    }
    else{
        if (root->left != NULL){
            root->left = addOneRow((root->left), val, depth-1);
        }
        if (root->right != NULL){
            root->right = addOneRow((root->right), val, depth-1);
        }
    }
    return root;
}
